csrc/ir/nodes.cpp

```diff
@@ -389,9 +389,64 @@ UnaryOp::UnaryOp(IrBuilderPasskey passkey, UnaryOpType type, Val* out, Val* in)
 
 std::vector<PolymorphicValue> UnaryOp::evaluate(
     const ExpressionEvaluator& ee,
-    const std::vector<PolymorphicValue>& inputs) const {
+    std::unordered_map<const Val*, PolymorphicValue>& known_values) const {
   using namespace PolymorphicValue_functions;
-  const auto& in = inputs.at(0);
+
+  // If the UnaryOp is CastOp, check if the preceding pattern of
+  // operators matches with matmul (MmaOp(Broadcast (A), Broadcast(B)) -> Cast)
+  // or matmul + bias (BinaryOp::Add (MmaOp(Broadcast (A), Broadcast(B),
+  // Broadcast(bias)) -> Cast) If not, evaluate UnaryOp::CastOp along with the
+  // other types by evaluating the immediate input.
+
+  // Check if the unary op is a cast from fp32 to lower precision.
+  auto is_downcast = [this]() -> bool {
+    if (getUnaryOpType() != UnaryOpType::Cast) {
+      return false;
+    }
+    auto in_dtype = input(0)->getDataType().value();
+    return (
+        in_dtype == DataType::Float &&
+        isInclusiveType(*(out()->getDataType()), in_dtype));
+  };
+
+  if (is_downcast() && input(0)->definition() != nullptr) {
+    Val* mma_lhs = nullptr;
+    Val* mma_rhs = nullptr;
+
+    // Case 1: MmaOp + Cast (Matmul)
+    if (MmaOpUtils::matchMatmulCast(
+            this, std::ref(mma_lhs), std::ref(mma_rhs))) {
+      // Inputs to MmaOp are of the shape [M, K, 1] x [1, K, N]
+      const auto a =
+          ee.evaluate(mma_lhs, known_values).as<at::Tensor>().squeeze(-1);
+      const auto b =
+          ee.evaluate(mma_rhs, known_values).as<at::Tensor>().squeeze(0);
+      return {a.matmul(b)};
+    }
+
+    Val* bias = nullptr;
+    // Case 2: Matmul + Bias
+    if (MmaOpUtils::matchMatmulBiasCast(
```

Comment:
I am not sure how to address this here, but what worries me is torch.nn.functional.linear takes x.A(T)+bias. So unless we transpose A, we won't have [M,K,1] [1,N,K]. 

https://pytorch.org/docs/stable/generated/torch.nn.functional.linear.html

We may need to write/use a routine to detect the shape.

Reply:
I'll address this is in the next PR. The assumed shapes are the same as that in matmul, so I was assuming that the transpose would already have been done when needed. We can discuss more today.